### ISSUE TITLE
Numpy2ri submodule has a glitch, and misidentifies an object as 'ListSexpVector' 

### ISSUE THREAD

**Issue by ghayward** [rpy2/rpy2#890]
This is my first time ever reporting an issue to an official repo, so if I'm doing it incorrectly, please forgive me.

I detailed an issue over on StackOverflow and a user named [kesh](https://stackoverflow.com/users/4516027/kesh?tab=profile) solved it.

https://stackoverflow.com/questions/72524531/rpy2-vector-issue-with-listsexpvector-any-ideas-on-how-to-fix

This is what he said was the issue, and I wanted to bring it to your attention. 

> "I think you encountered a bug in the numpy2ri submodule. The issue you're encountering is that the local converter thinks hellow_world is ListSexpVector but it appears to be StrSexpVector object instead (based on my quick debugging). I don't know enough about numpy2ri module yet to know exactly how this is happening, but here is a workaround:"
[Comment link](https://stackoverflow.com/a/72525236/11736959).

***
And now I'll reproduce my code that caused the error.

tl;dr - any ideas of how to fix AttributeError: 'ListSexpVector' object has no attribute 'items'

Here's the full code. The imports for rpy2 may be different slightly for you:

```
import os
os.environ['R_HOME'] = "/Library/Frameworks/R.framework/Resources" 

import rpy2 
import rpy2.robjects as ro
from rpy2.robjects.packages import importr
from rpy2.robjects import pandas2ri, r, numpy2ri 
from rpy2.robjects.conversion import localconverter

pandas2ri.activate()
numpy2ri.activate()
```

Then I make a test dataframe.

```
import pandas as pd
data = {'status': ['treatment','treatment','control','control'], 
        'result': [20,21,19,18]}

df = pd.DataFrame(data)

```

Then I run the first test code, and it works.

```
with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter): 
   r.assign("pydf", df) 
   hellow_world = r("""
          summary(pydf)        
              """) 
   print(hellow_world)
```

I get the proper R print out. Woot. :)

But then when I try what I really want to do:

```
with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter): 
    r.assign("pydf", df) 
    hellow_world = r("""
          t.test(result ~ status, data = pydf)       
          """) 
    print(hellow_world)
```

I get the following error:

`AttributeError: 'ListSexpVector' object has no attribute 'items'`

It seems to have to do with the rpy2py method, but any thoughts on how to fix would be greatly appreciated!

Thank you. The Lord Always Delivers!

***

Then kesh solved it:

I think you encountered a bug in the numpy2ri submodule. The issue you're encountering is that the local converter thinks hellow_world is ListSexpVector but it appears to be StrSexpVector object instead (based on my quick debugging). I don't know enough about numpy2ri module yet to know exactly how this is happening, but here is a workaround:

If you just want to print the R message

```
r.assign("pydf", df) 
hellow_world = r("""
    t.test(result ~ status, data = pydf)       
    """) 
print(hellow_world)
```

You don't need a converter at all.

If you want to parse R message in Python

```
r.assign("pydf", df) 
hellow_world = r("""
    t.test(result ~ status, data = pydf)       
    """) 
with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter): 
    print(list(hellow_world))
```

Without the localconverter context hellow_world = r(... line returns a raw R object without triggering the erroneous conversion. Then, converting hellow_world to a Python list within the conversion context will trigger functional per-list-item conversions. The resulting list is a lot harder to read, but it is a Python list with str and numpy objects.

***

So I wanted to bring this to your attention, since there may be a glitch!

Best,
George
#TheLordAlwaysDelivers

**Comment by lgautier** [comment]
```python
import rpy2 
import rpy2.robjects as ro
from rpy2.robjects.packages import importr
from rpy2.robjects import pandas2ri, r, numpy2ri 
from rpy2.robjects.conversion import localconverter

pandas2ri.activate()
numpy2ri.activate()
```

The use of `activate()` is discouraged because it can result in cryptic error, and it is deprecated. A deprecation warning should happen when calling those methods. Moreover, you are using local converters later in your code. 

```python
with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter): 
    r.assign("pydf", df) 
    hellow_world = r("""
          t.test(result ~ status, data = pydf)       
          """) 
    print(hellow_world)
```
Here the pandas2ri conversion rules meant to replace numpy2ri conversion rules might be reverted ti numpy2ri converter rules. Pandas conversion rules include numpy conversion rules (and modify/augment them).

If doing the following there a conversion issue though (`RecursionError: maximum recursion depth exceeded in comparison`):

```python
with localconverter(ro.default_converter + pandas2ri.converter): 
    r.assign("pydf", df) 
    hellow_world = r("""
          t.test(result ~ status, data = pydf)       
          """) 
    print(hellow_world)
```

Switching back to the default converter can be a temporary workaround:
```python
with localconverter(ro.default_converter + pandas2ri.converter): 
    r.assign("pydf", df) 
    with localconverter(ro.default_converter):
        hellow_world = r("""
              t.test(result ~ status, data = pydf)       
              """) 
    print(hellow_world)
```

**Comment by lgautier** [comment]
Minimal way to reproduce:
```python
import rpy2.robjects as ro
from rpy2.robjects import pandas2ri
from rpy2.robjects.conversion import localconverter

with localconverter(ro.default_converter + pandas2ri.converter):
    foo = ro.r('list(a=1, b="c")')
```

**Comment by lgautier** [comment]
Different error with the numpy converter. Not sure about how independent both errors are:

```python
import rpy2.robjects as ro
from rpy2.robjects import numpy2ri
from rpy2.robjects.conversion import localconverter
with localconverter(ro.default_converter + numpy2ri.converter):
    foo = ro.r('list(a=1)')
```

**Comment by ghayward** [comment]
Thank you so much @lgautier for these detailed responses. I'm going through them now. I can't emphasize enough to the greater data science community how important `rpy2` is because it lets us integrate some statistical analysis into complex python scripts. The use cases are infinite. 

I never thought of 'double' looping, if you will, the `localconverter`, so I'll be checking that out. 

I also never knew how to beat the Recursion error, so thank you so much for that (`RecursionError: maximum recursion depth exceeded in comparison`).

I thought the `.activate()`'s were mandatory; very glad to know I should stop using them.

**Comment by lgautier** [comment]
Fixed.
